File: day-6/ruby/solution.rb
lines = File.open('../inputs/test.txt').readlines.map(&:chomp)

def get_marker_index(lines, packet_size)
    lines.map do |line|
        _, match_idx = line.chars
                            .each_cons(packet_size)
                            .with_index
                            .find { |str, i| str.uniq.size == packet_size }
        match_idx + packet_size
    end
end

puts "Markers ending at the following indices for packet size 4: #{get_marker_index(lines, 4).join(', ')}."
puts "Markers ending at the following indices for packet size 14: #{get_marker_index(lines, 14).join(', ')}."

File: day-6/python/solution.py
signals = open('../inputs/test.txt', 'r').read().splitlines()


def find_marker_end(signals, marker_size):
    indices = []
    for signal in signals:
        for i in range(len(signal)):
            if len(set(signal[i:i + marker_size])) == marker_size:
                indices.append(i + marker_size)
                break
    return indices


print("Signals start at indices: {}".format(find_marker_end(signals, 4)))
print("Signals start at indices: {}".format(find_marker_end(signals, 14)))
